File: Common/Core/Testing/Python/TestSetBuffer.py
"""Test SetBuffer and BufferChangedEvent for AOS and SOA data array templates.

Verifies that SetBuffer() correctly transfers buffer data, and that
BufferChangedEvent fires from SetBuffer, SetArray, and ShallowCopy.
"""

from vtkmodules.vtkCommonCore import (
    vtkCommand,
    vtkDoubleArray,
    vtkFloatArray,
    vtkSOADataArrayTemplate,
)
from vtkmodules.test import Testing


class TestSetBuffer(Testing.vtkTest):
    """Test SetBuffer and BufferChangedEvent for AOS and SOA arrays."""

    def _make_observer(self):
        """Create an observer callback that counts invocations."""
        state = {"count": 0}

        def callback(obj, event):
            state["count"] += 1

        return callback, state

    # ------------------------------------------------------------------
    # AOS SetBuffer
    # ------------------------------------------------------------------

    def testAOSSetBuffer(self):
        """AOS: SetBuffer transfers data correctly."""
        src = vtkFloatArray()
        src.SetNumberOfComponents(2)
        src.SetNumberOfTuples(3)
        for i in range(6):
            src.SetValue(i, float(i + 1))

        buf = src.GetBuffer()

        dst = vtkFloatArray()
        dst.SetNumberOfComponents(2)
        dst.SetBuffer(buf, True)

        self.assertEqual(dst.GetNumberOfTuples(), 3)
        for i in range(6):
            self.assertAlmostEqual(dst.GetValue(i), float(i + 1))

        t = dst.GetTuple(1)
        self.assertAlmostEqual(t[0], 3.0)
        self.assertAlmostEqual(t[1], 4.0)

    def testAOSSetBufferEvent(self):
        """AOS: SetBuffer fires BufferChangedEvent."""
        src = vtkFloatArray()
        src.SetNumberOfComponents(1)
        src.SetNumberOfTuples(4)

        dst = vtkFloatArray()
        dst.SetNumberOfComponents(1)

        cb, state = self._make_observer()
        dst.AddObserver(vtkCommand.BufferChangedEvent, cb)

        dst.SetBuffer(src.GetBuffer(), True)
        self.assertGreater(state["count"], 0,
                           "BufferChangedEvent should fire on AOS SetBuffer")

    def testAOSSetArrayEvent(self):
        """AOS: SetArray fires BufferChangedEvent."""
        arr = vtkFloatArray()
        arr.SetNumberOfComponents(1)

        cb, state = self._make_observer()
        arr.AddObserver(vtkCommand.BufferChangedEvent, cb)

        arr.SetNumberOfTuples(4)
        state["count"] = 0  # reset after allocation events

        other = vtkFloatArray()
        other.SetNumberOfComponents(1)
        other.SetNumberOfTuples(4)
        arr.SetArray(other, 4, True)

        self.assertGreater(state["count"], 0,
                           "BufferChangedEvent should fire on AOS SetArray")

    def testAOSShallowCopyEvent(self):
        """AOS: ShallowCopy fires BufferChangedEvent."""
        src = vtkFloatArray()
        src.SetNumberOfComponents(2)
        src.SetNumberOfTuples(3)

        dst = vtkFloatArray()
        dst.SetNumberOfComponents(2)

        cb, state = self._make_observer()
        dst.AddObserver(vtkCommand.BufferChangedEvent, cb)

        dst.ShallowCopy(src)
        self.assertGreater(state["count"], 0,
                           "BufferChangedEvent should fire on AOS ShallowCopy")

    # ------------------------------------------------------------------
    # SOA SetBuffer
    # ------------------------------------------------------------------

    def testSOASetBuffer(self):
        """SOA: SetBuffer transfers per-component data correctly."""
        numTuples = 4
        # Create source arrays to get buffers from
        comp0 = vtkDoubleArray()
        comp0.SetNumberOfComponents(1)
        comp0.SetNumberOfTuples(numTuples)
        for i in range(numTuples):
            comp0.SetValue(i, float(i))

        comp1 = vtkDoubleArray()
        comp1.SetNumberOfComponents(1)
        comp1.SetNumberOfTuples(numTuples)
        for i in range(numTuples):
            comp1.SetValue(i, float(10 + i))

        arr = vtkSOADataArrayTemplate['float64']()
        arr.SetNumberOfComponents(2)
        arr.SetBuffer(0, comp0.GetBuffer(), True)
        arr.SetBuffer(1, comp1.GetBuffer(), False)

        self.assertEqual(arr.GetNumberOfTuples(), numTuples)
        t = arr.GetTuple(2)
        self.assertAlmostEqual(t[0], 2.0)
        self.assertAlmostEqual(t[1], 12.0)

    def testSOASetBufferEvent(self):
        """SOA: SetBuffer fires BufferChangedEvent."""
        arr = vtkSOADataArrayTemplate['float64']()
        arr.SetNumberOfComponents(2)
        arr.SetNumberOfTuples(1)

        cb, state = self._make_observer()
        arr.AddObserver(vtkCommand.BufferChangedEvent, cb)

        comp = vtkDoubleArray()
        comp.SetNumberOfComponents(1)
        comp.SetNumberOfTuples(4)
        arr.SetBuffer(0, comp.GetBuffer(), True)

        self.assertGreater(state["count"], 0,
                           "BufferChangedEvent should fire on SOA SetBuffer")

    def testSOASetArrayEvent(self):
        """SOA: SetArray fires BufferChangedEvent."""
        arr = vtkSOADataArrayTemplate['float64']()
        arr.SetNumberOfComponents(2)
        arr.SetNumberOfTuples(1)

        cb, state = self._make_observer()
        arr.AddObserver(vtkCommand.BufferChangedEvent, cb)

        other = vtkDoubleArray()
        other.SetNumberOfComponents(1)
        other.SetNumberOfTuples(4)
        arr.SetArray(0, other, 4, False, True)

        self.assertGreater(state["count"], 0,
                           "BufferChangedEvent should fire on SOA SetArray")

    def testSOAShallowCopyEvent(self):
        """SOA: ShallowCopy fires BufferChangedEvent."""
        src = vtkSOADataArrayTemplate['float64']()
        src.SetNumberOfComponents(2)
        src.SetNumberOfTuples(3)

        dst = vtkSOADataArrayTemplate['float64']()
        dst.SetNumberOfComponents(2)

        cb, state = self._make_observer()
        dst.AddObserver(vtkCommand.BufferChangedEvent, cb)

        dst.ShallowCopy(src)
        self.assertGreater(state["count"], 0,
                           "BufferChangedEvent should fire on SOA ShallowCopy")


if __name__ == "__main__":
    Testing.main([(TestSetBuffer, 'test')])

File: Common/Core/Testing/Cxx/TestSetBuffer.cxx
// SPDX-FileCopyrightText: Copyright (c) Ken Martin, Will Schroeder, Bill Lorensen
// SPDX-License-Identifier: BSD-3-Clause
// Test SetBuffer() for AOS and SOA data array templates, and verify that
// BufferChangedEvent fires from SetBuffer, SetArray, and ShallowCopy.

#include "vtkAOSDataArrayTemplate.h"
#include "vtkBuffer.h"
#include "vtkCallbackCommand.h"
#include "vtkCommand.h"
#include "vtkNew.h"
#include "vtkSOADataArrayTemplate.h"

#include <iostream>

namespace
{

int EventCount = 0;

void OnBufferChanged(vtkObject*, unsigned long, void*, void*)
{
  EventCount++;
}

// Helper: attach a BufferChangedEvent observer and reset the counter.
vtkNew<vtkCallbackCommand> AttachObserver(vtkObject* obj)
{
  EventCount = 0;
  vtkNew<vtkCallbackCommand> cb;
  cb->SetCallback(OnBufferChanged);
  obj->AddObserver(vtkCommand::BufferChangedEvent, cb);
  return cb;
}

} // namespace

int TestSetBuffer(int, char*[])
{
  int retVal = 0;

  // ===================================================================
  // AOS SetBuffer — verify data is accessible
  // ===================================================================
  {
    vtkNew<vtkBuffer<double>> buffer;
    buffer->Allocate(6);
    double* ptr = buffer->GetBuffer();
    for (int i = 0; i < 6; i++)
    {
      ptr[i] = static_cast<double>(i + 1);
    }

    vtkNew<vtkAOSDataArrayTemplate<double>> arr;
    arr->SetNumberOfComponents(2);
    arr->SetBuffer(buffer, true);

    if (arr->GetNumberOfTuples() != 3)
    {
      std::cerr << "AOS SetBuffer: expected 3 tuples, got " << arr->GetNumberOfTuples() << "\n";
      retVal++;
    }
    for (int i = 0; i < 6; i++)
    {
      if (arr->GetValue(i) != static_cast<double>(i + 1))
      {
        std::cerr << "AOS SetBuffer: value mismatch at index " << i << "\n";
        retVal++;
        break;
      }
    }
    double tuple[2];
    arr->GetTuple(1, tuple);
    if (tuple[0] != 3.0 || tuple[1] != 4.0)
    {
      std::cerr << "AOS SetBuffer: GetTuple(1) returned wrong values\n";
      retVal++;
    }
  }

  // ===================================================================
  // AOS SetBuffer without updateMaxId
  // ===================================================================
  {
    vtkNew<vtkBuffer<double>> buffer;
    buffer->Allocate(10);

    vtkNew<vtkAOSDataArrayTemplate<double>> arr;
    arr->SetNumberOfComponents(1);
    arr->SetNumberOfTuples(0);
    arr->SetBuffer(buffer, false);

    if (arr->GetNumberOfTuples() != 0)
    {
      std::cerr << "AOS SetBuffer (no updateMaxId): expected 0 tuples, got "
                << arr->GetNumberOfTuples() << "\n";
      retVal++;
    }
  }

  // ===================================================================
  // AOS SetBuffer fires BufferChangedEvent
  // ===================================================================
  {
    vtkNew<vtkAOSDataArrayTemplate<double>> arr;
    arr->SetNumberOfComponents(1);
    AttachObserver(arr);

    vtkNew<vtkBuffer<double>> buffer;
    buffer->Allocate(4);
    arr->SetBuffer(buffer, true);

    if (EventCount < 1)
    {
      std::cerr << "AOS SetBuffer did not fire BufferChangedEvent\n";
      retVal++;
    }
  }

  // ===================================================================
  // AOS SetArray fires BufferChangedEvent
  // ===================================================================
  {
    vtkNew<vtkAOSDataArrayTemplate<double>> arr;
    arr->SetNumberOfComponents(1);
    AttachObserver(arr);

    double data[] = { 1.0, 2.0, 3.0 };
    arr->SetArray(data, 3, 1); // save=1 so array won't free

    if (EventCount < 1)
    {
      std::cerr << "AOS SetArray did not fire BufferChangedEvent\n";
      retVal++;
    }
  }

  // ===================================================================
  // AOS ShallowCopy fires BufferChangedEvent
  // ===================================================================
  {
    vtkNew<vtkAOSDataArrayTemplate<double>> src;
    src->SetNumberOfComponents(2);
    src->SetNumberOfTuples(3);
    for (int i = 0; i < 6; i++)
    {
      src->SetValue(i, static_cast<double>(i));
    }

    vtkNew<vtkAOSDataArrayTemplate<double>> dst;
    dst->SetNumberOfComponents(2);
    AttachObserver(dst);

    dst->ShallowCopy(src);

    if (EventCount < 1)
    {
      std::cerr << "AOS ShallowCopy did not fire BufferChangedEvent\n";
      retVal++;
    }
  }

  // ===================================================================
  // SOA SetBuffer — verify data is accessible
  // ===================================================================
  {
    const int numTuples = 4;
    vtkNew<vtkBuffer<double>> buf0;
    buf0->Allocate(numTuples);
    vtkNew<vtkBuffer<double>> buf1;
    buf1->Allocate(numTuples);

    for (int i = 0; i < numTuples; i++)
    {
      buf0->GetBuffer()[i] = static_cast<double>(i);
      buf1->GetBuffer()[i] = static_cast<double>(10 + i);
    }

    vtkNew<vtkSOADataArrayTemplate<double>> arr;
    arr->SetNumberOfComponents(2);
    arr->SetBuffer(0, buf0, true);
    arr->SetBuffer(1, buf1, false);

    if (arr->GetNumberOfTuples() != numTuples)
    {
      std::cerr << "SOA SetBuffer: expected " << numTuples << " tuples, got "
                << arr->GetNumberOfTuples() << "\n";
      retVal++;
    }
    double tuple[2];
    arr->GetTuple(2, tuple);
    if (tuple[0] != 2.0 || tuple[1] != 12.0)
    {
      std::cerr << "SOA SetBuffer: GetTuple(2) returned wrong values\n";
      retVal++;
    }
  }

  // ===================================================================
  // SOA SetBuffer fires BufferChangedEvent
  // ===================================================================
  {
    vtkNew<vtkSOADataArrayTemplate<double>> arr;
    arr->SetNumberOfComponents(2);
    arr->SetNumberOfTuples(1);
    AttachObserver(arr);

    vtkNew<vtkBuffer<double>> buffer;
    buffer->Allocate(4);
    arr->SetBuffer(0, buffer, true);

    if (EventCount < 1)
    {
      std::cerr << "SOA SetBuffer did not fire BufferChangedEvent\n";
      retVal++;
    }
  }

  // ===================================================================
  // SOA SetArray fires BufferChangedEvent
  // ===================================================================
  {
    vtkNew<vtkSOADataArrayTemplate<double>> arr;
    arr->SetNumberOfComponents(2);
    arr->SetNumberOfTuples(1);
    AttachObserver(arr);

    double data[] = { 1.0, 2.0, 3.0 };
    arr->SetArray(0, data, 3, false, true);

    if (EventCount < 1)
    {
      std::cerr << "SOA SetArray did not fire BufferChangedEvent\n";
      retVal++;
    }
  }

  // ===================================================================
  // SOA ShallowCopy fires BufferChangedEvent
  // ===================================================================
  {
    vtkNew<vtkSOADataArrayTemplate<double>> src;
    src->SetNumberOfComponents(2);
    src->SetNumberOfTuples(3);
    for (int i = 0; i < 3; i++)
    {
      double tuple[2] = { static_cast<double>(i), static_cast<double>(10 + i) };
      src->SetTuple(i, tuple);
    }

    vtkNew<vtkSOADataArrayTemplate<double>> dst;
    dst->SetNumberOfComponents(2);
    AttachObserver(dst);

    dst->ShallowCopy(src);

    if (EventCount < 1)
    {
      std::cerr << "SOA ShallowCopy did not fire BufferChangedEvent\n";
      retVal++;
    }
  }

  return retVal;
}
